src/neural-network.js

```diff
@@ -311,22 +304,17 @@ export default class NeuralNetwork {
    *    NOTE: Activation is stored directly on JSON object and not in the training options
    */
   _getTrainOptsJSON() {
-    let results = {}
-    if (this.trainOpts.iterations) { results.iterations = this.trainOpts.iterations; }
-    if (this.trainOpts.errorThresh) { results.errorThresh = this.trainOpts.errorThresh; }
-    if (this.trainOpts.logPeriod) { results.logPeriod = this.trainOpts.logPeriod; }
-    if (this.trainOpts.learningRate) { results.learningRate = this.trainOpts.learningRate; }
-    if (this.trainOpts.momentum) { results.momentum = this.trainOpts.momentum; }
-    if (this.trainOpts.callback) { results.callback = this.trainOpts.callback; }
-    if (this.trainOpts.callbackPeriod) { results.callbackPeriod = this.trainOpts.callbackPeriod; }
-    if (this.trainOpts.timeout) { results.timeout = this.trainOpts.timeout; }
-    if (this.trainOpts.log) { results.log = true; }
-    return results;
+    return Object.keys(NeuralNetwork.trainDefaults)
+      .reduce((opts, opt) => {
+        if (this.trainOpts[opt]) opts[opt] = this.trainOpts[opt];
```

Comment:
I don't think we want to add the json object of a callback to the results

Reply:
I agree with this

Reply:
What _should_ it be? If someone wants to output it via `JSON.stringify`, the keys with functions in them will get omitted. This is a little awkward, but sending in some string representation of the function means we'd have to account for this on both converting to and from a JSON.  If they instead are just using the JSON as a holder in-memory for their net's configuration (e.g. how the tests use it), it would actually provide meaningful data back with references to the actual functions used rather than a placeholder.

Reply:
Another question would be _why would we want to give the json of something someone else owns_?

Reply:
Said differently:  There are likely values we do want to retain as json, perhaps we could just be static in defining them, rather than including all of them.  If it were, for example, just a list of the ones that matter, that'd give us both the small footprint of code, and logic for what actually affected the net, which may be important later for json exporting.

Reply:
Following that train of thought, what are the values that actually affect the net?

Reply:
Also, @perkyguy you know how I feel about `reduce`.


Only sith lords deal in `reduce`

Reply:
@robertleeplummerjr join me, and together we can rule the galaxy!

Reply:
If we have a simple static list of the properties that affect the net that would later want to be referenced, it would be a static list and we'd simply use `map` on it.


p.s. @perkyguy I still see good in you, it is not too late

Reply:
so the `callback` and `log` are the only two options which now could be "owned" by someone else (`log` since https://github.com/BrainJS/brain.js/pull/137/files#diff-f0964eec5cfcdf6b9c5b85748849be07R336). should we trim both of these out? Or just one? Or transform them to `'function() { ... }'`, or do their `toString()`?

Reply:
So I am of the opinion that we probably don't want to include an unknown function in the json (or even console.log).  I think it would be perfectly acceptable to either

1) convert log to true when set and restore it as console.log
2) Ignore both the `log` and `callback` properties of the training

PS I also like Array.reduce:
![giphy](https://user-images.githubusercontent.com/2294139/35595293-6ae3ccc2-05d3-11e8-84dd-bc3a3f5a7cf0.gif)

Reply:
One thing we don't address is validation to the trainOpts themselves.  We _should probably_ validate each of the values assigned.  Imagine learningRates in the negatives, or basically anything that is bad.

Reply:
nm, I see it now

Reply:
My thought: using `reduce` is overkill.  Lets get a list of the items that should be specifically exported, then `map` them.

Reply:
it isn't that I hate `reduce`, it has it's place (dark side, unstructured data) and here we know the context and structure.

Reply:
so are you proposing we have `get exportableTrainingOptions()` which will give us a subset of `get trainingDefaults`?

Reply:
My initial thought was that this concept interfered with the general output would somehow make the library too non-specific enough for the lib, but the more I think about this, the more it makes sense.

ty guys for thinking this through, and it does clean up the training area.